contracts/Nonon/NononFriendCard.sol

```diff
@@ -0,0 +1,336 @@
+            )
+        );
+    }
+
+    function getSpriteSubstring(bytes memory spritesSvg, uint16 spriteIndex, uint16 spriteLength)
+        internal
+        pure
+        returns (bytes memory)
+    {
+        bytes memory sprite = new bytes(spriteLength);
+
+        for (uint256 i = 0; i < sprite.length; i++) {
+            sprite[i] = spritesSvg[i + spriteIndex];
+        }
+
+        return sprite;
+    }
+
+    // get metadata for token display based on a given points value
+    function levelData(uint256 tokenPoints) internal view returns (LevelImageData memory levelImageData) {
+        for (uint256 i = levels.length; i > 0;) {
+            Level memory level = levels[i - 1];
+            if (tokenPoints >= level.minimum) {
+                if (i < levels.length) {
+                    // there is at least one level above current, so get its minimum
+                    Level memory nextLevel = levels[i];
+                    return LevelImageData(
+                        level.name, level.colorGradient, level.spriteIndex, level.spriteLength, nextLevel.minimum
+                    );
+                } else {
+                    // highest level
+                    uint256 maxPoints = IERC721A(collectionAddress).totalSupply() * 2;
+                    return LevelImageData(level.name, level.colorGradient, level.spriteIndex, level.spriteLength, maxPoints);
+                }
+            }
+            unchecked {
+                --i;
+            }
+        }
+    }
+
+    // prevent transfer (except mint and burn)
+    function _beforeTokenTransfers(address from, address to, uint256, uint256) internal pure override {
+        if (from != address(0) && to != address(0)) {
+            revert OnlyForYou();
+        }
+    }
+
+    // add ID for associated sequential tokens to appropriate lists
+    function registerTokenMovement(address from, address to, uint256 collectionTokenStartId, uint256 quantity)
+        external
+        override
+        onlyCollection
+    {
+        if (from != address(0)) {
+            if (to != from) {
+                sentBitmap[from].setBatch(collectionTokenStartId, quantity);
+            }
+        }
+
+        if (to != address(0)) {
+            receivedBitmap[to].setBatch(collectionTokenStartId, quantity);
+        }
+
+        emit BatchMetadataUpdate(1, type(uint256).max);
+    }
+
+    // total points accumulated by a holder
+    function points(uint256 tokenId) public view returns (uint256) {
+        address owner = ownerOf(tokenId);
+        uint256 max = IERC721A(collectionAddress).totalSupply() + 1;
+
+        return receivedBitmap[owner].popCount(1, max) + sentBitmap[owner].popCount(1, max);
+    }
+
+    // convenience function to get point information in a token range
+    // note that this is expensive and most likely will require multiple calls to cover large ranges.
+    function tokenPointsInRange(uint256 startId, uint256 endId) public view returns (TokenPoints[] memory) {
```

Comment:
Is this used internally? or should it be an external method?

Reply:
changed to external in https://github.com/guminc/contracts/pull/73/commits/67285f429a95c1766f6f0e11e1d2c44eb1eeb991